Introducing Logical Operators for dataset conditional logic

sunank200 opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
We've expanded the dataset dependency handling in Airflow, building on PR apache/airflow#37016. This PR introduces the use of logical operators (`|` for OR and `&` for AND) to link `Dataset` instances, simplifying the expression of complex dependencies.

Key Enhancements:
- Logical operators `|` and `&` for combining `Dataset` objects.
- Enhanced URI validation within the `Dataset` class.
- New `DatasetAny` and `DatasetAll` classes for OR/AND conditions.
- `DatasetsExpression` class for building dataset condition trees.
- `extract_datasets` function to interpret these expression trees.

Example:
```python
dataset1 = Dataset(uri="s3://bucket1/data1")
dataset2 = Dataset(uri="s3://bucket2/data2")
dataset3 = Dataset(uri="s3://bucket3/data3")
dataset4 = Dataset(uri="s3://bucket4/data4")
dataset5 = Dataset(uri="s3://bucket5/data5")

expr = dataset1 | (dataset2 & dataset3)
# expr translates to DatasetAny(dataset1, DatasetAll(dataset2, dataset3)))

expr1 =  ((dataset1 & dataset2) | dataset3) & (dataset4 | dataset5)

# expr1 translates to DatasetAll(DatasetAny(DatasetAll(dataset1, dataset2), dataset3), DatasetAny(dataset4, dataset5))
```


This update offers a more intuitive way of expressing dataset dependencies in Airflow workflows.

_Depends on the merge of PR apache/airflow#37016._

## Dependency Checklist
- [x] PR #37016 should be merged before this PR.


---


<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

uranusjr commented on airflow/models/dag.py:
```diff
@@ -585,6 +587,8 @@ def __init__(
+        if isinstance(schedule, DatasetsExpression):
+            self.schedule = extract_datasets(dataset_expression=schedule)
         if isinstance(schedule, (DatasetAll, DatasetAny)):
             self.dataset_triggers = schedule
```
I wonder if we could unify those three into one base class. This doesn’t need to be done now.

sunank200 replied:
Sure I can do this as part of separate PR

sunank200 committed:
Implement | and & operators so that they can be used instead of DatasetAll and DatasetAny

Implement | and & operators so that they can be used instead of DatasetAll and DatasetAny

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -341,9 +341,9 @@ def __repr__(self) -> str:
-    Base class for boolean logic for dataset triggers.
+    Base class for boolean conditions on datasets. This class is intended for internal use only.
 
-    :meta private:
```
If it's for internal use only, why do we remove `:meta private`

sunank200 replied:
code is changed after rebase and this taken care of.

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -341,9 +341,9 @@ def __repr__(self) -> str:
+    Base class for boolean conditions on datasets. This class is intended for internal use only.
 
-    :meta private:
+    :param objects: A variable number of Dataset, DatasetAny, or DatasetAll instances.
```
If we already know the object types to expect, should we annotate the one in `__init__`

sunank200 replied:
code is changed after rebase and this taken care of.

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
 
     agg_func = any
 
+    def __init__(self, *objects: Dataset | DatasetAny | DatasetAll):
```
```suggestion
    def __init__(self, *objects: Dataset | DatasetAny | DatasetAll) -> None:
```

sunank200 replied:
changed.

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+        """Initialize with one or more Dataset, DatasetAny, or DatasetAll instances."""
+        super().__init__(*objects)
+
+    def __or__(self, other):
```
```suggestion
    def __or__(self, other) -> DatasetAny:
```

sunank200 replied:
Changed it.

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+            return DatasetAny(*self.objects, other)
+        return NotImplemented
+
+    def __and__(self, other):
```
```suggestion
    def __and__(self, other) -> DatasetAll:
```

sunank200 replied:
Changed it.

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+        return NotImplemented
+
+    def __repr__(self) -> str:
+        return f"DatasetAnd({', '.join(map(str, self.objects))})"
```
```suggestion
        return f"DatasetAll({', '.join(map(str, self.objects))})"
```

I guess it should be `DatasetAll` instead?

sunank200 replied:
Changed it

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+        """Initialize with one or more Dataset, DatasetAny, or DatasetAll instances."""
+        super().__init__(*objects)
+
+    def __or__(self, other):
```
```suggestion
    def __or__(self, other) -> DatasetAny:
```

sunank200 replied:
Changed it.

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+            return DatasetAny(self, other)
+        return NotImplemented
+
+    def __and__(self, other):
```
```suggestion
    def __and__(self, other) -> DatasetAll:
```

sunank200 replied:
Changed it.

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+    :param right: The right child node.
+    """
+
+    def __init__(self, value, left=None, right=None):
```
```suggestion
    def __init__(self, value, left=None, right=None) -> None:
```

sunank200 replied:
Changed it

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+    def __and__(self, other: Dataset | DatasetsExpression) -> DatasetsExpression:
+        return DatasetsExpression("&", self, other)
+
+    def __repr__(self):
```
```suggestion
    def __repr__(self) -> str:
```

sunank200 replied:
Changed it

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+            return repr(DatasetAll(self.left, self.right))
+        elif self.value == "|":
+            return repr(DatasetAny(self.left, self.right))
+
```
Do we need to handle `else`?

sunank200 replied:
Added else part now

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+            return repr(DatasetAny(self.left, self.right))
+
+
+def extract_datasets(dataset_expression: DatasetsExpression | Dataset):
```
```suggestion
def extract_datasets(dataset_expression: DatasetsExpression | Dataset) -> DatasetBooleanCondition | DatasetsExpression:
```

sunank200 replied:
Changed it

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+        elif dataset_expression.value == "|":
+            return DatasetAny(dataset_expression.left, dataset_expression.right)
+        else:
+            raise ValueError("Invalid Expression node value")
```
```suggestion

        raise ValueError("Invalid Expression node value")
```

sunank200 replied:
Changed it

Lee-W commented on airflow/models/dataset.py:
```diff
@@ -374,12 +386,96 @@ def all_datasets(self) -> dict[str, Dataset]:
+        else:
+            raise ValueError("Invalid Expression node value")
+    else:
+        return dataset_expression
```
```suggestion
    return dataset_expression
```

sunank200 replied:
Changed it

Lee-W commented on airflow/serialization/serialized_objects.py:
```diff
@@ -407,6 +407,8 @@ def serialize_to_json(
             elif key == "dataset_triggers":
                 serialized_object[key] = cls.serialize(value)
+            elif key == "dataset_triggers":
+                serialized_object[key] = cls.serialize(value)
```
```suggestion
            elif key == "dataset_triggers":
                serialized_object[key] = cls.serialize(value)
```

sunank200 replied:
Changed it. I think it was messed during rebased

Lee-W commented on tests/serialization/test_dag_serialization.py:
```diff
@@ -554,6 +554,9 @@ def validate_deserialized_dag(self, serialized_dag, dag):
         }
+        compare_serialization_list = {
+            "dataset_triggers",
+        }
```
```suggestion
        compare_serialization_list = {
            "dataset_triggers",
        }
```

sunank200 replied:
Changed it.I think is was messed during rebase.

Lee-W commented on airflow/datasets/__init__.py:
```diff
@@ -55,3 +55,13 @@ def __eq__(self, other):
         return hash(self.uri)
+
+    def __or__(self, other):
+        from airflow.models.dataset import DatasetAny
```
May I know why do we want to import here? Is there anything expensive in `airflow.models.dataset`

sunank200 replied:
It was to deal with cyclic imports

sunank200 replied:
The code is changed  now

Lee-W commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+
+To schedule a DAG to run only when two specific datasets have both been updated, use the AND operator (``&``):
+
+.. code-block:: python
```
Should we use `exampleinclude` instead?

sunank200 replied:
I just followed other examples in the doc. Other example uses `code-block:: python`

Lee-W commented on tests/models/test_datasets.py:
```diff
@@ -0,0 +1,87 @@
+from airflow.models.dataset import Dataset, DatasetAll, DatasetAny, extract_datasets
+
+
+def datasets_equal(d1, d2):
```
Should we implement `__eq__` in `Dataset`, `DatasetAny` `DatasetAll` instead?

sunank200 replied:
This is just required for the test currently.

sunank200 committed:
Refactor dataset class inheritance (#37590)

* Refactor DatasetAll and DatasetAny inheritance

They are moved from airflow.models.datasets to airflow.datasets since
the intention is to use them with Dataset, not DatasetModel. It is more
natural for users to import from the latter module instead.

A new (abstract) base class is added for the two classes, plus the OG
Dataset class, to inherit from. This allows us to replace a few
isinstance checks with simple molymorphism and make the logic a bit
simpler.

---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
Co-authored-by: Wei Lee <weilee.rx@gmail.com>

sunank200 committed:
Refactor the dataset inheritance

sunank200 committed:
Merge branch 'main' into add-dataset-conditional-syntax

jedcunningham commented on airflow/example_dags/example_datasets.py:
```diff
@@ -15,26 +15,40 @@
+dataset dependencies. This DAG is configured to execute either when both dataset_produces_1 and
+dataset_produces_2 datasets have been updated or according to a specific cron schedule, showcasing
+Airflow's versatility in handling mixed triggers for dataset and time-based scheduling.
+
```
```suggestion
```

nit: extra newline

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+        dag_id="consume_1_and_2_with_dataset_expressions",
+        start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
+        schedule=(dag1_dataset & dag2_dataset),
+    ) as dag:
```
```suggestion
    ):
```

nit

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+        dag_id="consume_1_or_2_with_dataset_expressions",
+        start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
+        schedule=(dag1_dataset | dag2_dataset),
+    ) as dag:
```
```suggestion
    ):
```

nit

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+    from airflow.operators.bash import BashOperator
+    from airflow.datasets import Dataset
+    import pendulum
+
```
My 2c, these don't have to be fully functional. I think it's easier if we just have the parts that are critical to the topic at hand.

sunank200 replied:
Sure. Changed it.

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+    with DAG(
+        dag_id="consume_1_and_2_with_dataset_expressions",
+        start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
+        schedule=(dag1_dataset & dag2_dataset),
```
```suggestion
        schedule=(dag1_dataset & dag2_dataset),
        ...
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+    with DAG(
+        dag_id="consume_1_or_2_with_dataset_expressions",
+        start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
+        schedule=(dag1_dataset | dag2_dataset),
```
```suggestion
        schedule=(dag1_dataset | dag2_dataset),
        ...
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+            task_id="consume_1_or_2_with_dataset_expressions",
+            bash_command="sleep 5",
+            outlets=[Dataset("s3://consuming_2_task/dataset_other_unknown.txt")],
+        )
```
```suggestion
        ...
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+            task_id="consume_1_and_2_with_dataset_expressions",
+            bash_command="sleep 5",
+            outlets=[Dataset("s3://consuming_2_task/dataset_other_unknown.txt")],
+        )
```
```suggestion
        ...
```

This is just the consumer, we don't need the downstream tasks.

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+    with DAG(
+        dag_id="consume_1_or_both_2_and_3_with_dataset_expressions",
+        start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
+        schedule=(dag1_dataset | (dag2_dataset & dag3_dataset)),
```
```suggestion
        schedule=(dag1_dataset | (dag2_dataset & dag3_dataset)),
        ...
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+        dag_id="consume_1_or_both_2_and_3_with_dataset_expressions",
+        start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
+        schedule=(dag1_dataset | (dag2_dataset & dag3_dataset)),
+    ) as dag:
```
```suggestion
    ):
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+            task_id="consume_1_or_both_2_and_3_with_dataset_expressions",
+            bash_command="sleep 5",
+            outlets=[Dataset("s3://consuming_2_task/dataset_other_unknown.txt")],
+        )
```
```suggestion
        ...
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -245,3 +324,5 @@ DatasetTimetable Integration
 
 For more detailed information on ``DatasetTimetable`` and its usage, refer to the corresponding section in :ref:`DatasetTimetable <dataset-timetable-section>`.
+
+These examples illustrate how Airflow's conditional dataset expressions can be used to create complex, data-dependent scheduling scenarios, providing precise control over when DAGs are triggered in response to data updates.
```
```suggestion
These examples illustrate how Airflow's conditional dataset expressions can be used to create complex data-dependent scheduling scenarios, providing precise control over when DAGs are triggered in response to data updates.
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+    dag2_dataset = Dataset("s3://dag2/output_1.txt")
+
+    with DAG(
+        dag_id="consume_1_and_2_with_dataset_expressions",
```
```suggestion
        # Consume dataset 1 and 2 with dataset expressions
```

A comment is probably even better than an example DAG id.

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+.. code-block:: python
+
+    with DAG(
+        dag_id="consume_1_or_2_with_dataset_expressions",
```
```suggestion
        # Consume dataset 1 or 2 with dataset expressions
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -237,6 +237,85 @@ Example:
+    dag3_dataset = Dataset("s3://dag3/output_3.txt")
+
+    with DAG(
+        dag_id="consume_1_or_both_2_and_3_with_dataset_expressions",
```
```suggestion
        # Consume dataset 1 or both 2 and 3 with dataset expressions
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/timetable.rst:
```diff
@@ -212,9 +212,29 @@ Here's an example of a DAG using ``DatasetTimetable``:
+    from airflow.timetables.trigger import CronTriggerTimetable
+
+    with DAG(
+        dag_id="conditional_dataset_and_time_based_timetable",
```
```suggestion
        # Conditional dataset and time based timetable
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/timetable.rst:
```diff
@@ -212,9 +212,29 @@ Here's an example of a DAG using ``DatasetTimetable``:
+        schedule=DatasetOrTimeSchedule(
+            timetable=CronTriggerTimetable("0 1 * * 3", timezone="UTC"), datasets=(dag1_dataset & dag2_dataset)
+        ),
+        tags=["dataset-time-based-timetable"],
```
```suggestion
        schedule=DatasetOrTimeSchedule(
            timetable=CronTriggerTimetable("0 1 * * 3", timezone="UTC"), datasets=(dag1_dataset & dag2_dataset)
        ),
        ...
```

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/timetable.rst:
```diff
@@ -212,9 +212,29 @@ Here's an example of a DAG using ``DatasetTimetable``:
+    from airflow.timetables import DatasetOrTimeSchedule
+    from airflow.timetables.trigger import CronTriggerTimetable
+
+    with DAG(
```
The rest of the examples on this page use taskflow, we should probably convert to that style too.

sunank200 replied:
Changed it

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/timetable.rst:
```diff
@@ -212,9 +212,29 @@ Here's an example of a DAG using ``DatasetTimetable``:
+            task_id="conditional_dataset_and_time_based_timetable",
+            bash_command="sleep 5",
+            outlets=[Dataset("s3://dataset_time_based/dataset_other_unknown.txt")],
+        )
```
```suggestion
        ...
```

jedcunningham commented on tests/datasets/test_dataset.py:
```diff
@@ -269,3 +269,69 @@ def test_dag_with_complex_dataset_triggers(session, dag_maker):
+        lambda: (dataset1 & dataset2) & (dataset3 & dataset4),
+        DatasetAll(dataset1, dataset2, DatasetAll(dataset3, dataset4)),
+    ),
+    (lambda: dataset1 | dataset2 | dataset3, DatasetAny(dataset1, dataset2, dataset3)),
```
```suggestion
    (lambda: dataset1 | dataset2 | dataset3, DatasetAny(dataset1, dataset2, dataset3)),
    (lambda: dataset1 & dataset2 & dataset3, DatasetAll(dataset1, dataset2, dataset3)),
```

Probably worth having coverage for and also.

jedcunningham commented on docs/apache-airflow/authoring-and-scheduling/timetable.rst:
```diff
@@ -212,9 +212,29 @@ Here's an example of a DAG using ``DatasetTimetable``:
+            timetable=CronTriggerTimetable("0 1 * * 3", timezone="UTC"), datasets=(dag1_dataset & dag2_dataset)
+        ),
+        tags=["dataset-time-based-timetable"],
+    ) as dag:
```
```suggestion
    ):
```

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -73,6 +73,12 @@ def __eq__(self, other):
     def __hash__(self):
         return hash(self.uri)
 
+    def __or__(self, other: Dataset):
```
```suggestion
    def __or__(self, other: Dataset | DatasetAny | DatasetAll) -> DatasetAny:
```

Can't other be anything?

sunank200 replied:
Changed it to `BaseDatasetEventInput`

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
+        """Initialize with one or more Dataset, DatasetAny, or DatasetAll instances."""
+        super().__init__(*objects)
+
+    def __or__(self, other):
```
```suggestion
    def __or__(self, other: Dataset | DatasetAny | DatasetAll) -> DatasetAny:
```

dstandish replied:
or perhaps `BaseDatasetEventInput` would work?

uranusjr replied:
Yes that’s why I added that base protocol

sunank200 replied:
Changed it to `BaseDatasetEventInput`

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -73,6 +73,12 @@ def __eq__(self, other):
+    def __or__(self, other: Dataset):
+        return DatasetAny(self, other)
+
+    def __and__(self, other: Dataset):
```
```suggestion
    def __and__(self, other: Dataset | DatasetAny | DatasetAll) -> DatasetAll:
```

sunank200 replied:
Changed it to `BaseDatasetEventInput`

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
+            return DatasetAny(*self.objects, other)
+        return NotImplemented
+
+    def __and__(self, other):
```
```suggestion
    def __and__(self, other: Dataset | DatasetAny | DatasetAll) -> DatasetAll:
```

uranusjr replied:
Best to use `BaseDatasetEventInput` here (and in `__and__` etc)

sunank200 replied:
Changed it to `BaseDatasetEventInput`

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
+            return DatasetAny(self, other)
+        return NotImplemented
+
+    def __and__(self, other):
```
```suggestion
    def __and__(self, other: Dataset  | DatasetAny | DatasetAll) -> DatasetAll:
```

sunank200 replied:
Changed it to `BaseDatasetEventInput`

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
+        """Initialize with one or more Dataset, DatasetAny, or DatasetAll instances."""
+        super().__init__(*objects)
+
+    def __or__(self, other):
```
```suggestion
    def __or__(self, other: Dataset | DatasetAny | DatasetAll) -> DatasetAny:
```

sunank200 replied:
Changed it to `BaseDatasetEventInput`

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
+
+    def __and__(self, other):
+        if isinstance(other, (Dataset, DatasetAny, DatasetAll)):
+            return DatasetAll(self, other)
```
Why is `or` using `*self.objects`, but this is using `self`? (Maybe that missing test will highlight this?)

sunank200 replied:
I think the current logic holds. Please correct me if I am wrong @jedcunningham. The `__or__` and `__and__` methods in the `DatasetAny` and `DatasetAll` classes are based on the intended semantics of "any" (logical OR) and "all" (logical AND) in the context of dataset dependencies.

### DatasetAny Class
- **__or__ Method**: When using the `__or__` method, if one of the operands is already a `DatasetAny` instance, we want to combine all its contained datasets with the new one, effectively extending the logical OR condition. Hence, `*self.objects, other` is correct because it takes all the existing datasets in the `DatasetAny` instance and combines them with the `other` dataset or condition, maintaining the OR relationship across all of them.

- **__and__ Method**: Conversely, when using the `__and__` method on a `DatasetAny`, combining it with another dataset or condition (regardless of its type) means we are transitioning from a logical OR to a logical AND relationship. This is a fundamental shift in logic: we're saying the new condition requires all previous conditions (represented by `self` as an existing `DatasetAny` instance) AND the new one to be true. Hence, `DatasetAll(self, other)` is correct because it constructs a new AND condition that groups all conditions encapsulated by `self` (which were previously in an OR relationship) with the `other`.

### DatasetAll Class
- **__or__ Method**: Similarly, for the `DatasetAll` class, when we use `__or__`, we are shifting from an AND relationship to an OR relationship. If `self` is a `DatasetAll` instance, it represents a set of datasets that were all required (AND condition). By using `DatasetAny(self, other)`, we create a new condition where either all datasets in `self` or the `other` dataset (or condition) need to be true, effectively transitioning from AND to OR.

- **__and__ Method**: The `__and__` method in `DatasetAll` is used to extend the existing AND condition with a new dataset or condition. Thus, `*self.objects, other` is correct because it adds the `other` condition to the existing set of conditions already within `DatasetAll`, preserving the AND relationship across all combined datasets.

WDYT? I will add the test for this

sunank200 replied:
Added the tests here: https://github.com/apache/airflow/pull/37101/commits/afc20df63cf7d263491282982a8eb020e33d8f9a

jedcunningham replied:
No you're right, looking at this again it makes perfect sense. I think I got confused we were in DatasetAll here, not DatasetAny.

jedcunningham commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
+        elif dataset_expression.value == "|":
+            return DatasetAny(dataset_expression.left, dataset_expression.right)
+        raise ValueError("Invalid Expression node value")
+    return dataset_expression
```
```suggestion
    if not isinstance(dataset_expression, DatasetsExpression):
        return dataset_expression

    if dataset_expression.value == "&":
        return DatasetAll(dataset_expression.left, dataset_expression.right)
    elif dataset_expression.value == "|":
        return DatasetAny(dataset_expression.left, dataset_expression.right)
    raise ValueError("Invalid Expression node value")
```

Bikeshedding a bit, but this removes some nesting.

dstandish commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
+        self.right = right
+
+    def __or__(self, other: Dataset | DatasetsExpression) -> DatasetsExpression:
+        return DatasetsExpression("|", self, other)
```
it does feel a little bit awkward to use the strings `|` and `&`  as values to signal how to handle.  maybe there's a better way to accomplish using subclasses?

uranusjr replied:
An Enum would be better.

sunank200 replied:
I have removed the `DatasetExpression` and using use Dataset, DatasetAll, and DatasetAny directly now as @uranusjr suggested

dstandish commented on airflow/jobs/scheduler_job_runner.py:
the changes to this file were in a previous version of my foundation PR, checking one at a time for exists.  i ultimately reverted that change before merging because i wasn't sure we needed it. i would keep that out of this PR cus it's unrelated.  we can consider it independently.  so i would revert the changes to th is file.

sunank200 replied:
I have [reverted](https://github.com/apache/airflow/pull/37101/commits/f40f6cd88194f2c99830643daefca80e85c2c7fb) it, Daniel. I had rebased on top of your PR initially.

sunank200 committed:
Apply suggestions from code review

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

sunank200 committed:
Fix documentation changes as per PR comments

Lee-W commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,91 @@ class DatasetAny(_DatasetBooleanCondition):
+        elif self.value == "|":
+            return repr(DatasetAny(self.left, self.right))
+        else:
+            return f"Invalid DatasetsExpression(value={self.value})"
```
```suggestion
        return f"Invalid DatasetsExpression(value={self.value})"
```

Not sure whether we should raise an exception in this case

sunank200 replied:
changed it

Lee-W commented on tests/datasets/test_dataset.py:
```diff
@@ -269,3 +269,70 @@ def test_dag_with_complex_dataset_triggers(session, dag_maker):
+@pytest.mark.parametrize("expression, expected", test_cases)
+def test_extract_datasets(expression, expected):
+    expr = expression()
+    result = extract_datasets(expr)
```
May I know why do we want to use `lambda` and function call here instead of using expression directly?

sunank200 replied:
Lambda functions are used to delay the evaluation of an expression until it is called. This means that the dataset objects and their combinations are not created until the test function is actually executed.

Lee-W replied:
Got it. Thanks!

Lee-W commented on tests/datasets/test_dataset.py:
```diff
@@ -269,3 +269,70 @@ def test_dag_with_complex_dataset_triggers(session, dag_maker):
     ), "Serialized 'dataset_triggers' should be a dict"
+
+
+def datasets_equal(d1, d2):
```
```suggestion
def datasets_equal(d1: Dataset | DatasetAny | DatasetAll, d2: Dataset | DatasetAny | DatasetAll) -> bool:
```

sunank200 replied:
Changed

sunank200 committed:
fix the typing and PR comments

sunank200 committed:
revert back the scheduler changes

sunank200 committed:
use Dataset, DatasetAll, and DatasetAny directly for dataset expression

sunank200 committed:
add the test for Dataset expressions

sunank200 committed:
Merge branch 'main' into add-dataset-conditional-syntax

uranusjr committed:
Remove some unneeded checks

uranusjr reviewed (approved):
I added a commit to put `__and__` and `__or__` in the base class instead, and add comment on the optimized `__and__` and `__or__` for clarity to avoid future readers needing to ask https://github.com/apache/airflow/pull/37101#discussion_r1499629521 again.

uranusjr committed:
Remove unneeded __init__

dstandish commented on airflow/datasets/__init__.py:
```diff
@@ -106,8 +112,22 @@ class DatasetAny(_DatasetBooleanCondition):
     agg_func = any
 
+    def __or__(self, other: BaseDatasetEventInput) -> DatasetAny:
+        # Optimization: X | (Y | Z) is equivalent to X | Y | Z.
```
@sunank200 i think you meant to mark this as a todo?

i.e.
```
# todo: Optimization: X | (Y | Z) is equivalent to X | Y | Z.
```

i.e. should just tweak the language to make clear this is a _potential_ optimization, not one that is already committed that you are simply telling us about

dstandish replied:
alternatively if you don't want to think of it as a "to do" you could write
```
# potential optimization: X | (Y | Z) is equivalent to X | Y | Z.
```

sunank200 replied:
I think this was last minute commit from @uranusjr  before merge: https://github.com/apache/airflow/pull/37101/commits/2d670847d74790eee482c2f684a8ab0e0ddb0bec I can do the change as required.

sunank200 replied:
TP had commented more context to this change here: https://github.com/apache/airflow/pull/37101#pullrequestreview-1900009726

uranusjr replied:
I messed up the example though, the thing we’re optimising here is `(X | Y) | Z` instead. Let’s fix this.
